LobbyServer2/BridgeServer/BridgeServerProtocol.cs

```diff
@@ -825,5 +824,162 @@ public void UpdateModsAndCatalysts()
+        {
+            if (playerInfo.CharacterType == CharacterType.PendingWillFill) return "";
+
+            var thiefChars = playerInfo.TeamId == Team.TeamA ? duplicateCharsA : duplicateCharsB;
+            return thiefChars.Where(p => p.CharacterType == playerInfo.CharacterType && p.AccountId != playerInfo.AccountId).Select(p => p.Handle).FirstOrDefault();
+        }
+
+        public void CheckIfAllSelected()
+        {
+            var teamACharacters = GetCharactersByTeam(Team.TeamA);
+            var teamBCharacters = GetCharactersByTeam(Team.TeamB);
+
+            var duplicateCharsA = GetDuplicateCharacters(teamACharacters);
+            var duplicateCharsB = GetDuplicateCharacters(teamBCharacters);
+            
+            foreach (long player in GetPlayers())
+            {
+                LobbyServerPlayerInfo playerInfo = GetPlayerInfo(player);
+                LobbyServerProtocol playerConnection = SessionManager.GetClientConnection(player);
+                PersistedAccountData account = DB.Get().AccountDao.GetAccount(playerInfo.AccountId);
+
+                if (playerConnection != null && IsCharacterUnavailable(playerInfo, duplicateCharsA, duplicateCharsB) && playerInfo.ReadyState != ReadyState.Ready)
+                {
+                    CharacterType randomType = account.AccountComponent.LastCharacter;
+                    if (account.AccountComponent.LastCharacter == playerInfo.CharacterType)
+                    {
+                        // If they do not press ready and do not select a new character force them a random character else use the one they selected
+                        randomType = AssignRandomCharacter(playerInfo, teamACharacters, teamBCharacters);
+                    }
+                    
+                    UpdateAccountCharacter(playerInfo, randomType);
+                    NotifyCharacterChange(playerConnection, playerInfo, randomType);
+                    SetPlayerReady(playerConnection, playerInfo, randomType);
+                }
+            }
+        }
+
+        private CharacterType AssignRandomCharacter(LobbyServerPlayerInfo playerInfo, ILookup<CharacterType, LobbyServerPlayerInfo> teamACharacters, ILookup<CharacterType, LobbyServerPlayerInfo> teamBCharacters)
+        {
+            Random rand = new Random();
+            List<CharacterType> availableTypes = CharacterConfigs.Characters
+                .Where(cc => cc.Value.AllowForPlayers && cc.Value.CharacterRole != CharacterRole.None)
+                .Select(cc => cc.Key)
+                .ToList();
+
+            CharacterType randomType = CharacterType.None;
+
+            while (true)
+            {
+                randomType = availableTypes[rand.Next(availableTypes.Count)];
+
+                if (playerInfo.TeamId == Team.TeamA && !teamACharacters[randomType].Any()) break;
+                if (playerInfo.TeamId == Team.TeamB && !teamBCharacters[randomType].Any()) break;
+            }
```

Comment:
Just filter used characters out along with restricted ones

Reply:
Can't really since team a can use the ones in team b and other way around so :)

Reply:
How do you avoid testing aganist the other team it in the loop then?

Reply:
It checks if they are in team a or b in while then uses there specific team achar... Or teambchar if char doesn't match use that char then, notice team*a* or team*b*

Reply:
Why do you even pass both teams if you only need the team of the player you are selecting a character for?

Reply:
Sure. Could just pass one team or move availabletypes outside function pass it and filter but does it matter? Or change it to foreach you can refactor it multiple times at best while loop runs 4times idk if you want to change it you can